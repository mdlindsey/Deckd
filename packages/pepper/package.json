{
  "name": "@taga/pepper",
  "version": "1.0.0",
  "description": "Pepper game engine and bot logic for the TagaJS library",
  "main": "dist/main.js",
  "scripts": {
    "test": "jest"
  },
  "author": "Dan Lindsey",
  "license": "MIT",
  "dependencies": {},
  "devDependencies": {}
}
