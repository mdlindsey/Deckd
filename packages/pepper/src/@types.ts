// Outward facing GameModel
export interface GameModel {
    readonly data: RoundData[]
    readonly round: RoundModel
    readonly state: ActionState
    readonly score: ScoreModel
    readonly rounds: RoundModel[]
    interact(action: ActionInput): GameModel // Exposed publicly
    denormalizeRounds(): void // Populates this.round
    reduceExpectedState(): void // Populates this.state
}
// ActionState dictates next move expectations
export interface ActionState {
    id: number
    player: number
    modifier: number // eg: on action PLAY modifier is suitId
}
// ActionInput is input by user for next move
export interface ActionInput {
    id: number
    player: number
    payload: number // eg: card being played
}
// Raw data (from db)
export interface RoundData {
    hands: number[][]
    actions: ActionData[]
}
export interface ActionData {
    id: number
    payload: number
}
// Modeling
export type ScoreModel = { tricks: number, score: number }[];
export interface RoundModel {
    bids: number[]
    swaps: number[]
    trump: number
    plays: number[]
    hands: number[][]
}