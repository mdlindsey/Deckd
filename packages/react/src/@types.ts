export interface CardProps {
    suit?: string
    face?: string
    disabled?: boolean
    onClick?: CardCallback
    onMouseOver?: CardCallback
}
export interface CardWrapperProps extends CardProps {
    children?: any
    className?: string
    onClick?: (event:any) => void
    onMouseOver?: (event:any) => void
}
export type CardCallback = (event?:any, card?:{suit:string, face:string}) => void;
