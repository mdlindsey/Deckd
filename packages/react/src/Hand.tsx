import styled, { css } from 'styled-components';

export default styled.div`
    display: inline-block;
    width: ${(props:{children?:any}) => props.children.length * 2 + 10}em;
    height: 18em;
    position: relative;
    left: ${(props:{children?:any}) => props.children.length * 0.2}em;
    -moz-transform: rotate(10deg);
    -webkit-transform: rotate(10deg);
    -o-transform: rotate(10deg);
    transform: rotate(1.75deg);
    .card {
        position: absolute;
        cursor: pointer;
        img {
            max-width: 10.675rem;
            max-height: 15rem;
        }
        :hover {
            margin-top: -2.5em;
            padding-bottom: 2.5em;
        }
    }
    .card.disabled {
        cursor: default;
        :hover {
            margin-top: 0;
            padding-bottom: 0;
        }
    }
    ${
        (props:{children?:any}) => {
            let styles = '';
            const offsetLeft = 2;
            const middle = (props.children.length - 1) / 2;
            for(let i = 0; i < props.children.length; i++) {
                // middle cards should be highest so get the min distance from either end for offset
                const style = `translate(${(i - middle) * 0.25}rem, ${Math.abs(i - middle) * 0.15}rem) rotate(${(i - middle) * 3}deg)`;
                styles += `.card:nth-of-type(${i+1}) {
                    left: ${i * offsetLeft}rem;
                    -moz-transform: ${style};
                    -webkit-transform: ${style};
                    -o-transform: ${style};
                    transform: ${style};
                }`
            }
            return css`${styles}`;
        }
    }
`;
