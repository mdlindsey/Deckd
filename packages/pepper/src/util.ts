import {
    SJ,
    HJ,
    CJ,
    DJ,
    SUIT_S,
    SUIT_H,
    SUIT_C,
    SUIT_D,
    MAX_BID,
    MIN_BID,
    ACTION_BID,
    ACTION_TRUMP,
    ACTION_SWAP,
    ACTION_PLAY,
    REQ_PLAYERS,
    REQ_CARDS_PER_SUIT,
    actionMap,
    suitMap,
    cardMap,
    REQ_CARDS_PER_PLAYER,
} from './config';
import { ActionState, ActionInput } from './@types';
export const deal = ():number[][] => {
    const hands:number[][] = [];
    for(let i = 0; i < REQ_PLAYERS; i++) {
        hands[i] = [];
        for(let j = 0; j < REQ_CARDS_PER_PLAYER; j++) {
            let randomCard = -1;
            while(randomCard < 0 || [].concat.apply([], hands).includes(randomCard)) {
                randomCard = Math.floor(Math.random() * Math.floor(REQ_PLAYERS * REQ_CARDS_PER_PLAYER));
            }
            hands[i].push(randomCard);
        }
    }
    return hands;
};
export const cardSuit = (card:number, trump:number):number => {
    if (isBowerTrump(card,trump)) {
        switch(card) {
            case SJ: return SUIT_C;
            case HJ: return SUIT_D;
            case CJ: return SUIT_S;
            case DJ: return SUIT_H;
        }
    }
    return Math.floor(card / REQ_CARDS_PER_SUIT);
};
export const isHighTrump = (card:number, trump:number):boolean => {
    return (
        (trump === SUIT_S && card === SJ) ||
        (trump === SUIT_H && card === HJ) ||
        (trump === SUIT_C && card === CJ) ||
        (trump === SUIT_D && card === DJ)
    );
};
export const isBowerTrump = (card:number, trump:number):boolean => {
    return (
        (trump === SUIT_S && card === CJ) ||
        (trump === SUIT_H && card === DJ) ||
        (trump === SUIT_C && card === SJ) ||
        (trump === SUIT_D && card === HJ)
    );
};

export const sortTrick = (cards:number[], trump:number):number[] => {
    const suitToFollow = cardSuit(cards[0], trump);
    return cards.sort((c1, c2) => {
        const s1 = cardSuit(c1, trump);
        const s2 = cardSuit(c2, trump);
        if (s1 === suitToFollow && s2 === suitToFollow) {
            return compareSuitedCards(c1, c2, trump);
        }
        if (s1 === suitToFollow && s2 !== suitToFollow) {
            return -1;
        }
        if (s1 !== suitToFollow && s2 === suitToFollow) {
            return 1;
        }
        // neither are the correct suit so it doesn't matter
        return -1;
    });
};
export const sortSuitedCards = (suitedCars:number[], trump:number):number[] => {
    return suitedCars.sort((c1, c2) => compareSuitedCards(c1, c2, trump));
};
export const compareSuitedCards = (c1:number, c2:number, trump:number):number => {
    const s1 = cardSuit(c1, trump);
    const s2 = cardSuit(c2, trump);
    if (s1 !== s2) {
        throw new Error('encountered unsuited cars in compareSuitedCards');
    }
    // see if its the trump suit
    if (s1 === trump) {
        // both cards are trump, determine which is higher
        if (isHighTrump(c1, trump)) {
            return -1;
        }
        if (isHighTrump(c2, trump)) {
            return 1;
        }
        if (isBowerTrump(c1, trump)) {
            return -1;
        }
        if (isBowerTrump(c2, trump)) {
            return 1;
        }
    }
    // no other conditions met so it's numerical comparison
    return c1 < c2 ? -1 : 1;
};

export const canFollowSuit = (suitToFollow:number, hand:number[], plays:number[], trump:number):boolean => {
    for(const card of hand) {
        if (!plays.includes(card) && cardSuit(card, trump) === suitToFollow) {
            return true;
        }
    }
    return false;
};
export const totalActivePlayers = (bids:number[]) => Math.max(...bids) !== MAX_BID ? REQ_PLAYERS : REQ_PLAYERS - 1;
export const activeTrick = (bids:number[], plays:number[]):number[] => {
    const trickStartingIndex = plays.length < totalActivePlayers(bids) ? 0 : plays.length - (plays.length % totalActivePlayers(bids));
    return plays.slice(trickStartingIndex);
};
export const prevTrick = (bids:number[], plays:number[]):number[] => {
    const trickStartingIndex = Math.floor(plays.length / totalActivePlayers(bids)) * totalActivePlayers(bids);
    return plays.slice(trickStartingIndex - totalActivePlayers(bids), trickStartingIndex + totalActivePlayers(bids));
};

export const cardOwner = (card:number, hands:number[][]):number => {
    for(const playerIndex in hands) {
        if (hands[playerIndex].includes(card)) {
            return Number(playerIndex);
        }
    }
    return -1;
};

export const trickTakers = (bids:number[], hands:number[][], plays:number[], trump:number):number[] => {
    // Group plays into tricks
    const tricks:number[][] = [];
    const playerCount = totalActivePlayers(bids);
    for(let i = 0; i < plays.length; i++) {
        const trickIndex = Math.floor(i / playerCount);
        if (!tricks[trickIndex]) {
            tricks[trickIndex] = [];
        }
        tricks[trickIndex].push(plays[i]);
    } // probably have a bug with tricktaker being -1 from no owner
    // Loop through tricks and determine who took each one
    const takers:number[] = [];
    for(let i = 0; i < tricks.length; i++) {
        const rankedTrick = sortTrick(tricks[i], trump);
        takers[i] = cardOwner(rankedTrick[0], hands);
    }
    // Determine who took the last trick
    return takers;
};

export const translateAction = (action:ActionState|ActionInput|any):string => {
    if (action.modifier !== undefined) {
        // Dealing with action state
        return translateActionState(action);
    }
    return translateActionInput(action);
};

const translateActionState = (action:ActionState):string => {
    const verb = action.id === ACTION_TRUMP ? 'choose trump' : actionMap[action.id];
    let modifier = '';
    if (action.id === ACTION_BID) {
        modifier = `at least ${action.modifier < MIN_BID ? MIN_BID : action.modifier}`;
    }
    if (action.id === ACTION_PLAY || action.id === ACTION_SWAP) {
        modifier = action.modifier === -1 ? 'any card' : `a ${suitMap[action.modifier]}`;
    }
    return `Player #${action.player+1} to ${verb} ${modifier}`;
};

const translateActionInput = (action:ActionInput):string => {
    let caption = '';
    switch(action.id) {
        case ACTION_TRUMP:
            caption = `chose ${action.payload < 0 ? 'no' : `${suitMap[action.payload]}s`} trump`;
            break;
        case ACTION_SWAP:
            caption = action.payload < 0 ? 'declined swap' : `swapped ${cardMap[action.payload]}`;
            break;
        case ACTION_PLAY:
            caption = `played ${cardMap[action.payload]}`;
            break;
        default:
            caption = `${actionMap[action.id]}s ${action.payload}`;
    }
    return `Player #${action.player+1} ${caption}`;
};

export const sortCardsForHand = (cards:number[], trump:number):number[] => {
    // Group by suit
    const suits:number[][] = [[],[],[],[]];
    for(const card of cards) {
        suits[cardSuit(card, trump)].push(card);
    }
    // Order suits highest to lowest
    for(const suit in suits) {
        suits[suit] = sortSuitedCards(suits[suit], trump);
    }
    return [].concat.apply([], suits.sort((s1:number[], s2:number[]) => s2.length - s1.length));
};
