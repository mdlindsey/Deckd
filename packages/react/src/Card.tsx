import React from 'react';
import styled from 'styled-components';
import { CardWrapperProps, CardProps } from './@types';
/*
    <Card />
    suit - letter of suit (eg: S)
    face - letter of face (eg: A, K, Q, 10=T)
    disabled - if truthy will dim the card and prevent uplift on hover for HandWrapper
*/
// Wrap the cards with the max dimensions of the smallest image (back.png)
const CardWrapper:any = styled.span`
    img {
        max-width: 427px;
        max-height: 600px;
    }
    ${
        (props:CardWrapperProps) => props.disabled && `
            ::after {
                content:'';
                display: block;
                position: absolute;
                top: 0;
                left: 0;
                z-index: 1;
                width: 100%;
                height: 100%;
                border-radius: 10px;
                background: rgba(0, 0, 0, .6);
            }
        `
    }
`;
export default ({ suit='', face='', disabled=false, onClick=()=>{}, onMouseOver=()=>{} }:CardProps) => {
    const cardImg = validSuit(suit) && validFace(face) ? String(suit + face).toUpperCase() : 'BACK';
    return (
        <CardWrapper
            disabled={disabled}
            onClick={(e:any) => onClick(e,{suit,face})} 
            onMouseOver={(e:any) => onMouseOver(e,{suit,face})}
            className={['card', disabled ? 'disabled' : ''].join(' ')}>
            <img src={ require(`../assets/cards/${cardImg}.png`) } alt={`${face}${suit}`} />
        </CardWrapper>
    );
};
const isStr = (str:any) => typeof str === 'string' || str instanceof String;
const validSuit = (suit:string) => {
    if (!isStr(suit))
        return false;
    switch(suit.toUpperCase()) {
        case 'S':
        case 'H':
        case 'C':
        case 'D':
            return true;
        default:
            return false;
    }
};
const validFace = (face:string) => {
    if (!isStr(face))
        return false;
    switch(face.toUpperCase()) {
        case 'A':
        case 'K':
        case 'Q':
        case 'J':
        case 'T':
        case '9':
        case '8':
        case '7':
        case '6':
        case '5':
        case '4':
        case '3':
        case '2':
            return true;
        default:
            return false;
    }
};
