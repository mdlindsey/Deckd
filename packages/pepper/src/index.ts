import * as Types from './@types';
import * as Util from './util';
import * as Config from './config';
import { GameInstance } from './game';
import { act } from './bot';

export default {
    Types,
    Util,
    Config,
    GameInstance,
    New: (roundData:Types.RoundData[]=[]) => {
        const game = new GameInstance(roundData);
        return {
            game,
            bot: () => act(game)
        };
    }
};