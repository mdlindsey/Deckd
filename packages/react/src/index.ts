import Card from './Card';
export { Card };

import Hand from './Hand';
export { Hand };