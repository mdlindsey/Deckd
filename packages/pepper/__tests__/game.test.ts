import 'ts-jest';
import Pepper from '../src';
import { hands } from './__mocks__/pepper';
import { SWAP_REFUSE } from '../src/config';

// ------------ Game Engine Tests ---------------- //

describe('Game Engine Tests', () => {
  test('Instantiation', () => {
    const pepper = Pepper.New();
    expect(pepper.game.state).toMatchSnapshot();
  });

  test('Denormalize: Player #1 | Bid | NA', () => {
    const pepper = Pepper.New([
      {
        hands: hands.pepper,
        actions: []
      }
    ]);
    expect(pepper.game.state.id).toBe(Pepper.Config.ACTION_BID);
    expect(pepper.game.state.player).toBe(0);
    expect(pepper.game.state.modifier).toBe(Pepper.Config.MOD_NA);
  });

  test('Denormalize: Bid | Pass | New Round', () => {
    const pepper = Pepper.New([
      {
        hands: hands.pepper,
        actions: [
          { id: Pepper.Config.ACTION_BID, payload: 0 },
          { id: Pepper.Config.ACTION_BID, payload: 0 },
          { id: Pepper.Config.ACTION_BID, payload: 0 },
          { id: Pepper.Config.ACTION_BID, payload: 0 },
        ]
      }
    ]);
    expect(pepper.game.state.id).toBe(Pepper.Config.ACTION_DEAL);
  });

  test('Interaction: Must follow suit', () => {
    const pepper = Pepper.New([
      {
        hands: hands.threeBid,
        actions: [
          { id: Pepper.Config.ACTION_BID, payload: 3 },
          { id: Pepper.Config.ACTION_BID, payload: 0 },
          { id: Pepper.Config.ACTION_BID, payload: 0 },
          { id: Pepper.Config.ACTION_BID, payload: 0 },
          { id: Pepper.Config.ACTION_TRUMP, payload: -1 },
          { id: Pepper.Config.ACTION_PLAY, payload: Pepper.Config.SA },
        ]
      }
    ]);
    const mustFollowSuit = () => pepper.game.interact({ id: Pepper.Config.ACTION_PLAY, player: 1, payload: Pepper.Config.HA });
    expect(mustFollowSuit).toThrow(Error);
  });

  test('Interaction: Excessive bid', () => {
    const pepper = Pepper.New([
      {
        hands: hands.pepper,
        actions: []
      }
    ]);
    const excessiveBid = () => pepper.game.interact({ id: Pepper.Config.ACTION_BID, player: 0, payload: 99 });
    expect(excessiveBid).toThrow(Error);
  });

  test('Interaction: Invalid trump suit', () => {
    const pepper = Pepper.New([
      {
        hands: hands.pepper,
        actions: [
          { id: Pepper.Config.ACTION_BID, payload: 3 },
          { id: Pepper.Config.ACTION_BID, payload: 0 },
          { id: Pepper.Config.ACTION_BID, payload: 0 },
          { id: Pepper.Config.ACTION_BID, payload: 0 },
        ]
      }
    ]);
    const invalidTrump = () => pepper.game.interact({ id: Pepper.Config.ACTION_TRUMP, player: 0, payload: 99 });
    expect(invalidTrump).toThrow(Error);
  });

  test('Interaction: Invalid swap refusal', () => {
    const pepper = Pepper.New([
      {
        hands: hands.pepper,
        actions: [
          { id: Pepper.Config.ACTION_BID, payload: 6 },
          { id: Pepper.Config.ACTION_BID, payload: 0 },
          { id: Pepper.Config.ACTION_BID, payload: 0 },
          { id: Pepper.Config.ACTION_BID, payload: 0 },
          { id: Pepper.Config.ACTION_TRUMP, payload: -1 },
          { id: Pepper.Config.ACTION_SWAP, payload: 5 },
        ]
      }
    ]);
    const invalidSwapRefusal = () => pepper.game.interact({ id: Pepper.Config.ACTION_SWAP, player: 2, payload: SWAP_REFUSE });
    expect(invalidSwapRefusal).toThrow(Error);

    const unownedCard = () => pepper.game.interact({ id: Pepper.Config.ACTION_SWAP, player: 2, payload: 0 });
    expect(unownedCard).toThrow(Error);
  });

  test('Interaction: Unexpected action', () => {
    const pepper = Pepper.New();
    const unexpectedAction = () => pepper.game.interact({ id: Pepper.Config.ACTION_PLAY, player: 1, payload: Pepper.Config.SA });
    expect(unexpectedAction).toThrow(Error);
  });
  
  test('Flow: Player #1 | Pepper | No Trump | Swap', () => {
    const pepper = Pepper.New([
      {
        hands: hands.pepper,
        actions: []
      }
    ]);
    pepper.game.interact({ id: Pepper.Config.ACTION_BID, player: 0, payload: Pepper.Config.MAX_BID });
    expect(pepper.game.state.id).toBe(Pepper.Config.ACTION_TRUMP);
    expect(pepper.game.state.player).toBe(0);
    expect(pepper.game.state.modifier).toBe(Pepper.Config.MOD_NA);

    pepper.game.interact({ id: Pepper.Config.ACTION_TRUMP, player: 0, payload: Pepper.Config.TRUMP_NONE });
    expect(pepper.game.state.id).toBe(Pepper.Config.ACTION_SWAP);
    expect(pepper.game.state.player).toBe(0);
    expect(pepper.game.state.modifier).toBe(Pepper.Config.MOD_NA);

    pepper.game.interact({ id: Pepper.Config.ACTION_SWAP, player: 0, payload: Pepper.Config.S9 });
    expect(pepper.game.state.id).toBe(Pepper.Config.ACTION_SWAP);
    expect(pepper.game.state.player).toBe(2);
    expect(pepper.game.state.modifier).toBe(Pepper.Config.MOD_NA);

    pepper.game.interact({ id: Pepper.Config.ACTION_SWAP, player: 2, payload: Pepper.Config.CA });
    expect(pepper.game.state.id).toBe(Pepper.Config.ACTION_PLAY);
    expect(pepper.game.state.player).toBe(0);
    expect(pepper.game.state.modifier).toBe(Pepper.Config.MOD_NA);

    pepper.game.interact({ id: Pepper.Config.ACTION_PLAY, player: 0, payload: Pepper.Config.SA });
    expect(pepper.game.state.id).toBe(Pepper.Config.ACTION_PLAY);
    expect(pepper.game.state.player).toBe(1);
    expect(pepper.game.state.modifier).toBe(Pepper.Config.SUIT_S);

    pepper.game.interact({ id: Pepper.Config.ACTION_PLAY, player: 1, payload: Pepper.Config.H9 });
    expect(pepper.game.state.id).toBe(Pepper.Config.ACTION_PLAY);
    expect(pepper.game.state.player).toBe(3);
    expect(pepper.game.state.modifier).toBe(Pepper.Config.SUIT_S);

    pepper.game.interact({ id: Pepper.Config.ACTION_PLAY, player: 3, payload: Pepper.Config.D9 });
    expect(pepper.game.state.id).toBe(Pepper.Config.ACTION_PLAY);
    expect(pepper.game.state.player).toBe(0);
    expect(pepper.game.state.modifier).toBe(Pepper.Config.MOD_NA);
  });  

  test('Flow: Player #1 | 3-Bid', () => {
    const pepper = Pepper.New([
      {
        hands: hands.pepper,
        actions: []
      }
    ]);

    pepper.game.interact({ id: Pepper.Config.ACTION_BID, player: 0, payload: 3 });
    expect(pepper.game.state.id).toBe(Pepper.Config.ACTION_BID);
    expect(pepper.game.state.player).toBe(1);
    expect(pepper.game.state.modifier).toBe(3);

    pepper.game.interact({ id: Pepper.Config.ACTION_BID, player: 1, payload: 3 });
    expect(pepper.game.state.id).toBe(Pepper.Config.ACTION_BID);
    expect(pepper.game.state.player).toBe(2);
    expect(pepper.game.state.modifier).toBe(3);

    pepper.game.interact({ id: Pepper.Config.ACTION_BID, player: 2, payload: 3 });
    expect(pepper.game.state.id).toBe(Pepper.Config.ACTION_BID);
    expect(pepper.game.state.player).toBe(3);
    expect(pepper.game.state.modifier).toBe(3);
  });
});
