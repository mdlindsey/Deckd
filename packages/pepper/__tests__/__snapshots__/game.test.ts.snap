// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game Engine Snapshot Tests Denormalization 1`] = `Object {}`;

exports[`Game Engine Tests Instantiation 1`] = `
Object {
  "id": -1,
  "modifier": -1,
  "player": -1,
}
`;
